src/primitives/iter.rs

```diff
@@ -0,0 +1,473 @@
+// SPDX-License-Identifier: MIT
+
+//! Iterator Adaptors.
+//!
+//! Iterator extension traits and blanket implementations to convert:
+//!
+//! - `BytesToFes`: An iterator over bytes to an iterator over field elements.
+//! - `FesToBytes`: An iterator over field elements to an iterator over bytes.
+//! - `Checksummed`: An iterator over field elements that appends the checksum.
+//!
+//! # Examples
+//!
+//! ```
+//! use bech32::{Bech32, ByteIterExt, Fe32IterExt, Fe32, Hrp};
+//!
+//! let data = [
+//!     0x75, 0x1e, 0x76, 0xe8, 0x19, 0x91, 0x96, 0xd4,
+//!     0x54, 0x94, 0x1c, 0x45, 0xd1, 0xb3, 0xa3, 0x23,
+//!     0xf1, 0x43, 0x3b, 0xd6,
+//! ];
+//!
+//! // Convert byte data to GF32 field elements.
+//! let fe_iter = data.iter().copied().bytes_to_fes();
+//!
+//! // Convert field elements back to bytes.
+//! let byte_iter = fe_iter.fes_to_bytes();
+//!
+//! # assert!(data.iter().copied().eq(byte_iter));
+//! ```
+
+use crate::primitives::checksum::{self, Checksum, PackedFe32};
+use crate::primitives::encode::Encoder;
+use crate::primitives::gf32::Fe32;
+use crate::primitives::hrp::Hrp;
+
+/// Extension trait for byte iterators which provides an adaptor to GF32 elements.
+pub trait ByteIterExt: Sized + Iterator<Item = u8> {
+    /// Adapts the byte iterator to output GF32 field elements instead.
+    ///
+    /// If the total number of bits is not a multiple of 5 we pad with 0s
+    fn bytes_to_fes(mut self) -> BytesToFes<Self> {
+        BytesToFes { last_byte: self.next(), bit_offset: 0, iter: self }
+    }
+
+    /// Adapts the byte iterator to encode the field elements into a bech32 address (after first
+    /// converting it to a field element iterator).
+    fn encoder<I, Ck>(self, hrp: &Hrp) -> Encoder<BytesToFes<Self>, Ck>
+    where
+        I: Iterator<Item = Fe32>,
+        Ck: Checksum,
+    {
+        self.bytes_to_fes().encoder(hrp)
+    }
+}
+
+impl<I> ByteIterExt for I where I: Iterator<Item = u8> {}
+
+/// Extension trait for field element iterators.
+pub trait Fe32IterExt: Sized + Iterator<Item = Fe32> {
+    /// Adapts the `Fe32` iterator to output bytes instead.
+    ///
+    /// If the total number of bits is not a multiple of 8, any trailing bits
+    /// are simply dropped.
+    fn fes_to_bytes(mut self) -> FesToBytes<Self> {
+        FesToBytes { last_fe: self.next(), bit_offset: 0, iter: self }
+    }
+
+    /// Adapts the Fe32 iterator to encode the field elements into a bech32 address.
+    fn encoder<Ck: Checksum>(self, hrp: &Hrp) -> Encoder<Self, Ck> { Encoder::new(self, hrp) }
```

Comment:
In 5c9b549a3738cec82f633c14e274854b82edce45:

I'd rename `encoder` to `with_checksum`

Reply:
I guess `with_checksum_hrp`. And we could add a `with_checksum` method which just creates a `Checksummed`, no HRP.

Reply:
I was trying to show from the name that its not an iterator adapter but a builder.

Reply:
I'd argue that it's still an iterator adaptor, because it's a method on an arbitrary iterator. I get your point that it's not an iterator by itself, but I think it's more confusing/less discoverable to give it a name like `encoder`, which could mean pretty-much anything.

Reply:
Fair points, I'll have another play with it.

Reply:
Changed the `encoder` function to be called `with_checksum` - FTR I don't really like it but I cannot come up with anything better or reasonable argument against your points :)